CoastalResilienceNetwork/tnc-network-site
# Create dev env, add Travis build job, and implement Backbone app

## Overview

This PR implements the TNC Network site from the prototypes. I used Backbone to structure the stateful components -- we load it, jQuery, and Underscore from a CDN to keep the footprint light. I also added a tentative `config.json` file which we can flesh out in subsequent work to configure the app on loading.

The commits show the arc of the work:

- first commit sets up a dev environment
- second adds the images, favicons, style rules, etc from the prototype to the `src` dir
- third implements the site as a Backbone app.

I'll add more details in the *Notes* field below.

Connects #3 
Connects #4 

## Demo

![tnc-network-site](https://user-images.githubusercontent.com/4165523/32065013-889be4be-ba49-11e7-95de-f2025e28b314.gif)

## Notes

*Dev env*

I adapted some of the dev setup from the Civic Apps repo, including an eslint setup, some project scripts for development, and also updated the README and added a .travis.yml file. The eslint setup is based on an ES5-friendly airbnb config which sort of matches our ES6/JSX projects. I subsequently disabled a few of the rules which didn't seem to match. The Travis job will build PRs & other branches.

*Implementation*

I kept the stateful stuff in Backbone. On load we hydrate the state from a config.json file, then track it and update as the state changes. I left non-configurable stuff -- like the modal -- mostly out of Backbone.

Currently all the Backbone elements get put on the `window.` scope. I considered trying requirejs or similar for module loading, but that'd introduce more overhead. One alternative we might choose is to just keep all of the Backbone stuff in a single long file rather than splitting it across App, models, and views.

*List animation*

There's one interaction I didn't touch: the animation from the prototype on selection a list item:

![tnc-prototype-animation](https://user-images.githubusercontent.com/4165523/32065378-9e458a94-ba4a-11e7-9cac-5572bf6f2495.gif)

I think we should handle this in a subsequent issue, and I left the `main.js` file so we'll have the code side-by-side.

## Testing
- get this branch, then `./scripts/setup` and `./scripts/server`
- visit the app on port 8642 and verify that it works and matches the prototype (minus the list animation noted above)
- run `./scripts/lint` and verify that it runs the linter
- check the README and verify that it's accurate wrt the project setup
